### DISCUSSION TITLE
Could we compile stylo/servo into web assembly easily?

### DISCUSSION THREAD

**Discussion by shiqimei** [servo/servo#28070] [Q&A] [answered]
Since servo/stylo is written in rust and I know rust supports wasm well. So my question is can we compile some components of this project (stylo for example) into wasm?

Facebook's open source layout engine [yoga](https://yogalayout.com/getting-started/standalone/) is mainly written in C++ and they officially provide pure JS compiled version cross-compiled using emscripten.

Does stylo have some heavy dependencies or other potential issues that might block us from compiling it into wasm?

PS. It will be cool if a web developer could use a full-featured style engine with JavaScript in the browser 🙈

**Comment by jdm** [comment, marked as answer]
The native dependencies like SpiderMonkey would cause problems. Similarly, all the code that currently depends on GL bindings like the WebGL implementation would interefere, and the networking code. In short, anything that is platform-specific in some way would likely need to be adjusted or rewritten to accomodate the new target. Also, it's not clear to me how easily threads translate to wasm right now, and Servo uses many threads. It would not be an easy project.

**Comment by WilstonOreo** [comment]
Do I understand this correctly? You want to compile Servo/Stylo to WASM so can run a browser engine inside another browser engine? Nice hack, but where is the use case?

**Reply by shiqimei** [reply]
There're some use cases already - [react-canvas](https://github.com/Flipboard/react-canvas). See this [article](https://engineering.flipboard.com/2015/02/mobile-web) for further information (in short, they tried to draw elements into canvas directly instead of DOM and use [yoga](https://yogalayout.com/getting-started/standalone/) for layout to achive 60 FPS+ in some performace sensitive cases).

I think they are on the right path but one possible reason that made this project dead might be **the lack of CSS support**. For web developers, we're already familiar with building the UI using css, which is the web standard. By compliling Style engine & Layout engine into wasm, we can propabaly draw some React components into canvas just like rendered in the DOM without any code changes but gain performance improvement. That's what my actual idea is.

**Reply by shiqimei** [reply]
> so can run a browser engine inside another browser engine? Nice hack, but where is the use case?

NOT the full engine (it's too HUGE) but only a small subset of it. Since React/Vue or some other frameworks already has their vDOM implementation. Maybe compliling stylo and a layout engine into wasm is enough.

**Reply by banekondic1996** [reply]
Use case. Use Google Maps without API :) . That's why I'm here at least. Google will be fixing it, as soon as someone does it, cause they earn from that

**Comment by johnhenry** [comment]
Checkout [shadow](https://github.com/canadahonk/shadow), an early stage, mostly javascript implementation of a rendering engine.
I wonder if the ecosystem has evolved such that it might now be possible to compile servo to wasm?

**Comment by nicoburns** [comment]
Stylo does compile to WASM with a small patch (https://github.com/servo/stylo/pull/78). However, I don't think anyone has tested whether it actually works there or not.
